peps/pep-9999.rst

```diff
@@ -0,0 +1,145 @@
+PEP: 9999
+Title: imath --- module for number-theoretic functions
+Author: Sergey B Kirpichev
+Sponsor: Victor Stinner <vstinner@python.org>
+Discussions-To: Pending
+Status: Draft
+Type: Standards Track
+Created: 10-May-2025
+Python-Version: 3.15
+Post-History: 02-Jun-2019,
+              `09-May-2025 <https://discuss.python.org/t/91337>`__,
+
+
+Abstract
+========
+
+This PEP proposes a new module for number-theoretic functions, like
+:external+py3.14:func:`math.gcd` or :external+py3.14:func:`math.isqrt`.
+
+
+Motivation
+==========
+
+The :external+py3.14:mod:`math` documentaton says: "This module provides access
+to the mathematical functions defined by the C standard."  But as a state of
+art, over time the module was populated with functions that aren't related to
+the C standard or floating-point arithmetics.  Now it's much harder to describe
+module scope, content and interfaces (e.g. returned values).
+
+For example, the :external+py3.14:mod:`math` module documentation says: "Except
+when explicitly noted otherwise, all return values are floats."  This is not
+longer true:  *None* of the functions listed in the "Number-theoretic
+functions" [1]_ subsection of the documentation returns a float, but the
+documentation doesn't say so.  In the proposed module a similar sentence:
+"Except when explicitly noted otherwise, all return values are integers." could
+tell the truth once.
+
+Apparently, the :external+py3.14:mod:`math` can't serve as a catch-all place
+for mathematical functions: we have also the :external+py3.14:mod:`cmath` and
+the :external+py3.14:mod:`statistics`.  Let's make same for integer-related
+functions.  It would provide shared context, which reduces verbosity in the
+documentation and conceptual load.  It also aids discoverability through
+grouping related functions and IDEs suggesting helpful completions.
+
+
+Specification
+=============
+
+The PEP proposes moving the following integer-related functions [1]_ in a new
+module, called ``imath``:
+
+* :external+py3.14:func:`~math.comb`
+* :external+py3.14:func:`~math.factorial`
+* :external+py3.14:func:`~math.gcd`
+* :external+py3.14:func:`~math.isqrt`
+* :external+py3.14:func:`~math.lcm`
+* :external+py3.14:func:`~math.perm`
+
+Their aliases in :external+py3.14:mod:`math` will be :term:`soft deprecated`.
+
+Possible extensions for the new module and it's scope are discussed in the
+`Open Issues <Open Issues_>`_ section.  New functions are not part of the
+proposal.
+
+
+Backwards Compatibility
+=======================
+
+As aliases in :external+py3.14:mod:`math` will be kept for indefinite time
+(their use would be discouraged), there are no anticipated code breaks.
+
+
+Reference Implementation
+========================
+
+TBD
+
+
+Open Issues
+===========
+
+Module name
+-----------
+
+There is already an ``imath`` project on PyPI, but only with two releases, with
+the most recent one 4 years ago.  Its repository is no longer accessible.
```

Comment:
the pypi project may be abandoned, but [this Imath](https://github.com/AcademySoftwareFoundation/Imath) is maintained and fairly active.
Naive code search across my employer's repo has hundreds of hits referring to *that* `imath`.
could you consider going with a different name please? :)

Reply:
I appreciate options.  Current version is a compromise, based on the [pool](https://discuss.python.org/t/split-out-integer-functions-from-the-math-module/91337/35).  As you could see, I'm in a tiny camp of "something else".  My initial proposal was "ntheory" (it also has a dead toy project on PyPI).

I would guess, that words like "random" or "fractions" have even more hits.  BTW, linked project seems unrelated to the Python ecosystem.

Reply:
> BTW, linked project seems unrelated to the Python ecosystem.

What do you mean by "unrelated"? It's a C++ library with Python bindings!

Reply:
Hmm, indeed.  Though, as naming is still in the "Open Issues", I wouldn't care too much right now.

@vstinner, what do you think if we start from the "ntheory" variant?